### PR TITLE
Modeler 2.4c: PortFieldSum and substitution

### PR DESCRIPTION
**Pull request by payetvin**

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+{
+    std::size_t operator()(const Nodes::PortFieldSumNode& n) const;
+};
+
+/**
+ * @brief Represents the context for performing substitutions in a syntax tree.
+ */
+struct PortFieldSumSubstitutionContext
+{
+    std::unordered_map<Nodes::PortFieldSumNode, Nodes::Node*, KeyHasher> portfield;
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.h]:
```suggestion
    std::unordered_map<Nodes::PortFieldSumNode, std::vector<Nodes::Node*>, KeyHasher> portfield;
```

File: src/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    CloneVisitor(registry),
+    ctx_(ctx)
+{
+}
+
+Nodes::Node* PortFieldSumSubstitutionVisitor::visit(const Nodes::PortFieldSumNode* node)
+{
+    if (auto it = ctx_.portfield.find(*node); it != ctx_.portfield.end())
+    {
+        return it->second;
```
Comment (flomnes) [inline src/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.cpp]: `it->second` is now a `std::vector<Node*>`, that can be provided as a parameter to `SumNode::SumNode`.

```suggestion
    return registry_.create<SumNode>(it->second);
```

**Comment by flomnes** [comment]
Possible improvement https://gist.github.com/flomnes/5ca35a28b8037e9e0ed7f84900cd7b8d

File: src/solver/expressions/hashable.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Hashable::Hashable(const std::string& s1, const std::string& s2):
+    s1(s1),
+    s2(s2)
+{
+}
+
+bool Hashable::operator==(const Hashable& other) const
+{
+    return s1 == other.s1 && s2 == other.s2;
+}
```
Comment (flomnes) [inline src/solver/expressions/hashable.cpp:37]:
```suggestion
bool Hashable::operator==(const Hashable& other) const = default;
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+{
+    std::size_t operator()(const Nodes::PortFieldSumNode& n) const;
+};
+
+/**
+ * @brief Represents the context for performing substitutions in a syntax tree.
+ */
+struct PortFieldSumSubstitutionContext
+{
+    std::unordered_map<Nodes::PortFieldSumNode, std::vector<Nodes::Node*>, PortFieldSumKeyHasher>
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.h]:
```suggestion
    std::unordered_map<Nodes::PortFieldSumNode, std::vector<Nodes::Node*>, PortFieldHash>
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+#include "antares/solver/expressions/visitors/CloneVisitor.h"
+
+namespace Antares::Solver::Visitors
+{
+
+struct PortFieldSumKeyHasher
+{
+    std::size_t operator()(const Nodes::PortFieldSumNode& n) const;
+};
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.h]:
```suggestion
```

File: src/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+std::size_t PortFieldSumKeyHasher::operator()(const Nodes::PortFieldSumNode& n) const
+{
+    std::size_t seed = 0;
+
+    boost::hash_combine(seed, boost::hash_value(n.getPortName()));
+    boost::hash_combine(seed, boost::hash_value(n.getFieldName()));
+
+    return seed;
+}
```
Comment (flomnes) [inline src/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.cpp]:
```suggestion
```

File: src/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#include <boost/functional/hash.hpp>
+
```
Comment (flomnes) [inline src/solver/expressions/visitors/PortFieldSumSubstitutionVisitor.cpp]:
```suggestion
```

File: src/tests/src/solver/expressions/test_SubstitutionVisitor.cpp [hunk 8]
```diff
@@ -113,10 +114,12 @@ class SubstitutionFixture: public Registry<Node>
 
     Node* substitute(Node* original)
     {
-        PortfieldSubstitutionContext ctx;
-        ctx.portfield.emplace(PortFieldNode("port", "literal"), create<LiteralNode>(10));
+        PortFieldSubstitutionContext ctx;
+        ctx.portfield.emplace(std::piecewise_construct,
```
Comment (a-zakir) [inline src/tests/src/solver/expressions/test_SubstitutionVisitor.cpp:118]: 🥇

**Comment by a-zakir** [comment]
~add test on~:

 
~~std::string name() const override~~
   ~~{~~
       ~~return "PortFieldSumNode";~~
  ~~}~~

File: src/tests/src/solver/expressions/test_AstDOTStyleVisitor.cpp [hunk 9]
```diff
@@ -135,7 +138,9 @@ legend_NegationNode -> legend_ParameterNode [style=invis];
 legend_ParameterNode [ label =" ParameterNode: 1"]
 legend_ParameterNode -> legend_PortFieldNode [style=invis];
 legend_PortFieldNode [ label =" PortFieldNode: 1"]
-legend_PortFieldNode -> legend_SubtractionNode [style=invis];
+legend_PortFieldNode -> legend_PortFieldSumNode [style=invis];
+legend_PortFieldSumNode [ label =" PortFieldSumNode: 1"]
+legend_PortFieldSumNode -> legend_SubtractionNode [style=invis];
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_AstDOTStyleVisitor.cpp:143]: 🥇

**Review by flomnes — APPROVED** [review]

**Pull request merged**